pkg/container-collection/container-collection.go

```diff
@@ -0,0 +1,213 @@
+// Copyright 2019-2021 The Inspektor Gadget authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package containercollection provides the ContainerCollection struct to keep
+// track of the set of running containers and primitives to query that set with
+// various criteria.
+//
+// It is used by the Gadget Tracer Manager to keep track of containers part of
+// Kubernetes pods and by Local Gadget Manager to keep track of containers on a
+// Linux system.
+package containercollection
+
+import (
+	"sync"
+
+	pb "github.com/kinvolk/inspektor-gadget/pkg/gadgettracermanager/api"
+	"github.com/kinvolk/inspektor-gadget/pkg/gadgettracermanager/pubsub"
+)
+
+// ContainerCollection holds a set of containers. It can be embedded as an
+// anonymous struct to help other structs implement the ContainerResolver
+// interface. For this reason, some methods are prefixed with 'Container' to
+// make clear that
+type ContainerCollection struct {
+	// Keys:   Container ID
+	// Values: pb.ContainerDefinition
+	containers sync.Map
+
+	// subs contains a list of subscribers of container events
+	pubsub *pubsub.GadgetPubSub
+}
+
+// ContainerCollectionInitialize initializes a ContainerCollection. It is
+// useful when ContainerCollection is embedded as an anonymous struct because
+// we don't use a contructor in that case.
+func (cc *ContainerCollection) ContainerCollectionInitialize() {
+	cc.pubsub = pubsub.NewGadgetPubSub()
+}
+
+// ContainerGet looks up a container by the container id and return it if
+// found, or return nil if not found. The ok result indicates whether value was found in the map.
+func (cc *ContainerCollection) ContainerGet(id string) (value *pb.ContainerDefinition) {
+	v, ok := cc.containers.Load(id)
+	if !ok {
+		return nil
+	}
+	containerDefinition := v.(pb.ContainerDefinition)
+	return &containerDefinition
+}
+
+// ContainerRemove removes a container from the collection.
+func (cc *ContainerCollection) ContainerRemove(id string) {
+	c := cc.ContainerGet(id)
+	if c == nil {
+		return
+	}
+	cc.containers.Delete(id)
+
+	cc.pubsub.Publish(pubsub.EVENT_TYPE_REMOVE_CONTAINER, *c)
```

Comment:
Nit: [LoadAndDelete](https://pkg.go.dev/sync#Map.LoadAndDelete) could help to simplify this code.